openlibrary/core/wikidata.py

```diff
@@ -0,0 +1,147 @@
+WIKIDATA_CACHE_TTL_DAYS = 30
+
+
+@dataclass
+class WikidataEntity:
+    """
+    This is the model of the api response from WikiData plus the updated field
+    https://www.wikidata.org/wiki/Wikidata:REST_API
+    """
+
+    id: str
+    type: str
+    labels: dict
+    descriptions: dict
+    aliases: dict
+    statements: dict
+    sitelinks: dict
+    updated: datetime  # This is when we fetched the data, not when the entity was changed in Wikidata
+
+    def description(self, language: str = 'en') -> str | None:
+        """If a description isn't available in the requested language default to English"""
+        return self.descriptions.get(language) or self.descriptions.get('en')
+
+    @classmethod
+    def from_dict(cls, response: dict, updated: datetime):
+        return cls(
+            id=response['id'],
+            type=response['type'],
+            labels=response['labels'],
+            descriptions=response['descriptions'],
+            aliases=response['aliases'],
+            statements=response['statements'],
+            sitelinks=response['sitelinks'],
+            updated=updated,
+        )
+
+    def as_api_response_str(self) -> str:
+        """
+        Transforms the dataclass a JSON string like we get from the Wikidata API.
+        This is used for staring the json in the database.
+        """
+        entity_dict = {
+            'id': self.id,
+            'type': self.type,
+            'labels': self.labels,
+            'descriptions': self.descriptions,
+            'aliases': self.aliases,
+            'statements': self.statements,
+            'sitelinks': self.sitelinks,
+        }
+        return json.dumps(entity_dict)
+
+
+def _cache_expired(entity: WikidataEntity) -> bool:
+    return days_since(entity.updated) > WIKIDATA_CACHE_TTL_DAYS
+
+
+def get_wikidata_entity(
+    QID: str, bust_cache: bool = False, fetch_missing: bool = False
+) -> WikidataEntity | None:
+    """
+    This only supports QIDs, if we want to support PIDs we need to use different endpoints
+    By default this will only use the cache (unless it is expired).
+    This is to avoid overwhelming Wikidata servers with requests from every visit to an author page.
+    bust_cache must be set to True if you want to fetch new items from Wikidata.
+    """
+    if bust_cache:
+        _get_from_web(QID)
+
+    if entity := _get_from_cache(QID):
+        if _cache_expired(entity):
+            return _get_from_web(QID)
+        return entity
+
+    if fetch_missing and not entity:
+        return _get_from_web(QID)
+
+    return None
+
+
+def _get_from_web(id: str) -> WikidataEntity | None:
+    response = requests.get(f'{WIKIDATA_API_URL}{id}')
+    if response.status_code == 200:
+        entity = WikidataEntity.from_dict(
+            response=response.json(), updated=datetime.now()
+        )
+        _add_to_cache(entity)
+        return entity
+    else:
+        return None
```

Comment:
We unfortunately don't have a best practices for handling cases like this, but adding a logger.error should let us see if these are happening in Sentry.

```suggestion
        logger.error(f'Wikidata Response: {response.status_code}`, data={'id': id})
        return None
```

Reply:
I didn't see any examples in the code adding data like that so I put it in a string for now.